### PR TITLE
crimson/os/seastore: move AsyncCleaner to EPM

### PR DESCRIPTION
**Pull request by cyx1231st**
Also refer to https://github.com/ceph/ceph/pull/47489#issuecomment-1208188283

The generic device tiering work would probably break down into the following smaller items:
* Introduce generic interface for AsyncCleaner in EPM.
* Decouple trimming and segment reclaiming implementations for reuse.
* Split implementations between RBCleaner and SegmentCleaner.
* Design and implement generic device-tiering logic in the EPM, based on generations.
* Related tests.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by cyx1231st** [comment]
@zhscn

**Comment by zhscn** [comment]
LGTM

**Review by xxhdx1985126 — APPROVED** [review]
> /home/jenkins-build/build/workspace/ceph-pull-requests/src/test/crimson/seastore/transaction_manager_test_state.h:106:7: error: use of undeclared identifier 'init'
      init();

The logics of the code looks good to me. Only the test code compilation error needs to be fixed.

**Comment by myoungwon** [comment]
I'm trying to understand why AsyncCleaner needs to be moved to EPM (because commits here seems like cleanup). In my understanding, EPM is in charge of data placement, and AsyncCleaner has a role of trim or reclaim for now. Moving AsyncCleaner to EPM means that EPM become a manager that manages data placement, cleaning job, and tiering?

**Comment by cyx1231st** [comment]
> I'm trying to understand why AsyncCleaner needs to be moved to EPM (because commits here seems like cleanup).

The only effective commit is the one having the same title with this PR: https://github.com/ceph/ceph/pull/47646/commits/cb86ceb3be0a3dc9a84602e711486f5407c2d099. All the others are cleanups that are related to the changed code.

> In my understanding, EPM is in charge of data placement, and AsyncCleaner has a role of trim or reclaim for now. Moving AsyncCleaner to EPM means that EPM become a manager that manages data placement, cleaning job, and tiering?

This is to extend EPM with tiered devices -- data-placement needs to decide the device tier first. The specific cleaning implementation should not be implemented in EPM, only the higer-level coordinations are required among the device tiers.

> AsyncCleaner has a role of trim or reclaim for now.

AsyncCleaner also has a role to commit the extent allocations, see `AsyncCleaner::mark_space_used()` and `mark_space_free()`. And it is responsible to allocate segments -- an analog to allocate blocks.

File: src/crimson/os/seastore/async_cleaner.cc [hunk 1]
```diff
@@ -1336,7 +1377,7 @@ AsyncCleaner::reserve_projected_usage(std::size_t projected_usage)
   if (disable_trim) {
     return seastar::now();
   }
-  ceph_assert(init_complete);
+  ceph_assert(state >= cleaner_state_t::READY);
```
Comment (athanatos) [inline src/crimson/os/seastore/async_cleaner.cc:1380]: Probably worth adding an is_ready() helper for this.

File: src/crimson/os/seastore/extent_placement_manager.h [hunk 2]
```diff
@@ -263,7 +256,52 @@ class ExtentPlacementManager {
...
+    assert(test_device->get_device_type() == device_type_t::SEGMENTED);
+    add_device(test_device);
+    set_primary_device(test_device);
+  }
+
+  bool check_usage() {
+    return cleaner->check_usage();
+  }
+
+  seastar::future<> run_until_halt() {
```
Comment (athanatos) [inline src/crimson/os/seastore/extent_placement_manager.h:293]: run_background_work_until_halt()?

File: src/crimson/os/seastore/extent_placement_manager.h [hunk 3]
```diff
@@ -263,7 +256,52 @@ class ExtentPlacementManager {
...
+    return cleaner->reserve_projected_usage(projected_usage);
+  }
+
+  void release_projected_usage(std::size_t projected_usage) {
+    return cleaner->release_projected_usage(projected_usage);
+  }
+
+  // Testing interfaces
+
+  void test_init_no_cleaner(Device *test_device) {
```
Comment (athanatos) [inline src/crimson/os/seastore/extent_placement_manager.h:283]: test_init_no_background()?

**Review by athanatos — APPROVED** [review]
LGTM other than naming nits.  I think this is the right direction to go.  Because EPM is responsible for the relationships between the different devices, it makes sense for it also to be responsible for performing background processes related to tiering/cleaning.

**Comment by myoungwon** [comment]
> This is to extend EPM with tiered devices -- data-placement needs to decide the device tier first. The specific cleaning implementation should not be implemented in EPM, only the higer-level coordinations are required among the device tiers.

I got it.

File: src/crimson/os/seastore/async_cleaner.cc (hunk 1 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/async_cleaner.cc]: done

File: src/crimson/os/seastore/extent_placement_manager.h (hunk 2 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/extent_placement_manager.h]: done

File: src/crimson/os/seastore/extent_placement_manager.h (hunk 3 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/extent_placement_manager.h]: done

**Pull request merged**